src/pcm.rs

```diff
@@ -0,0 +1,228 @@
+    LfVcore0,
+    LfVcore1,
+    Lpm0LdoVcore0,
+    Lpm0LdoVcore1,
+    Lpm0DcdcVcore0,
+    Lpm0DcdcVcore1,
+    Lpm0LfVcore0,
+    Lpm0LfVcore1,
+    Lpm3,
+}
+
+/// Builder object that configures Power Control Manager (PCM)
+pub struct PcmConfig <'a, S: State>{
+    periph: &'a pac::pcm::RegisterBlock,
+    _state: S,
+}
+
+impl <'a, S>PcmConfig<'a, S> where S: State{
+    /// Converts PCM into a fresh, unconfigured PCM builder object
+    pub fn new() -> PcmConfig<'a, PcmDefined> {
+
+        let pcm = unsafe { &*PCM::ptr() };
+
+        PcmConfig {
+            periph: pcm,
+            _state: PcmDefined,
+        }
+    }
+
+    #[inline]
+    fn wait_pcm(&self) {
+        while (self.periph.pcmctl1.read().bits() >> 8) & 0x01 != 0 {
+            unsafe{llvm_asm!("NOP")};
+        }
+
+        for _n in 1..50 {
+            unsafe{llvm_asm!("NOP")};
+        }
+    }
+
+    #[inline]
+
+    fn set_reg_mask(&self, value: u16, mask: u16) {
+
+    /// CSKEY
+    const CSKEY: u32 = 0x695A0000;
+
+        self.periph.pcmctl0.modify(|r, w| unsafe {
+            w.bits((r.bits() & mask as u32) | CSKEY | value as u32)
+        });
+
+        self.periph.pcmctl0.modify(|r, w| unsafe {
+            w.bits(r.bits() & !CSKEY)
+        });
+    }
+}
+
+impl <'a>PcmConfig<'a, PcmDefined> {
+
+        #[inline]
+        pub fn set_vcore(&mut self, source: VCoreSel) {
+
+            let mut source_state: VCoreSel;
+            source_state = self.get_vcore();
+
+            if source_state == source {
+
+            for _n in 1..50 {
+                unsafe{llvm_asm!("NOP")};
+            }
+                return;
+            }
+
+            while source_state != source {
+
+                if source_state == VCoreSel::DcdcVcore1 ||
+                   source_state == VCoreSel::LfVcore1 {
+                    source_state = VCoreSel::LdoVcore1;
+                } else if source_state == VCoreSel::DcdcVcore0 ||
+                   source_state == VCoreSel::LfVcore0 {
+                   source_state = VCoreSel::LdoVcore0;
+                } else if source_state == VCoreSel::LdoVcore1 &&
+                   source != VCoreSel::DcdcVcore1 && source != VCoreSel::LfVcore1 {
+                   source_state = VCoreSel::LdoVcore0;
+                } else if source_state == VCoreSel::LdoVcore0 &&
+                    source != VCoreSel::DcdcVcore0 && source != VCoreSel::LfVcore0 {
+                     source_state = VCoreSel::LdoVcore1;
+                } else {
+                    source_state = source;
+                }
+
+                self.set_vcore_inline(source_state);
+                source_state = self.get_vcore();
+            }
+        }
+
+        #[inline]
+        fn set_vcore_inline(&mut self, source: VCoreSel) {
+
+            let mut status: u32;
+
+            let amr_mask: u16 = 0xFFF0;
+
+            /* save the PRIMASK into 'status' */
+            unsafe { llvm_asm!("mrs $0, PRIMASK" : "=r" (status) : : : "volatile") };
```

Comment:
What does this asm block do and why do we need it?

Reply:
This block saves the interrupt table and disable the maskable interrupts to perform a task. Some register changes should be executed without interruptions. Rust interrupt::free(|_| { ... will replace this block.